File: sdk/src/prompt-runtime.test.ts
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { PromptRuntimeClient, type ResolvedPrompt } from './prompt-runtime.js';

const SAMPLE: ResolvedPrompt = {
  promptId: 'p-1',
  promptName: 'refund-agent',
  promptVersionId: 'v-1',
  versionNumber: 3,
  environment: 'production',
  content: 'You are a refund agent v3',
  variablesSchema: { type: 'object' },
  modelDefaults: { temperature: 0.2 },
  etag: '"pv-abc123"',
  deployedAt: '2026-08-26T00:00:00.000Z',
};

function jsonResponse(body: unknown, init: ResponseInit = {}) {
  return new Response(JSON.stringify(body), {
    status: 200,
    headers: { 'Content-Type': 'application/json', ...(init.headers as Record<string, string>) },
    ...init,
  });
}

function notModified(etag: string) {
  return new Response(null, { status: 304, headers: { ETag: etag } });
}

describe('PromptRuntimeClient', () => {
  let fetchMock: ReturnType<typeof vi.fn>;

  beforeEach(() => {
    fetchMock = vi.fn();
    (globalThis as unknown as { fetch: typeof fetch }).fetch = fetchMock as unknown as typeof fetch;
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  function makeClient() {
    return new PromptRuntimeClient({
      baseUrl: 'https://api.example.com',
      apiKey: 'amt_test',
      projectId: 'proj-1',
      defaultMaxAgeMs: 1000,
      fetchImpl: fetchMock as unknown as typeof fetch,
    });
  }

  it('首次请求调用 GET /runtime/prompts 并写入缓存', async () => {
    fetchMock.mockResolvedValueOnce(jsonResponse(SAMPLE, { headers: { ETag: SAMPLE.etag } }));
    const client = makeClient();

    const p = await client.get('refund-agent');
    expect(p.promptVersionId).toBe('v-1');
    expect(p.versionNumber).toBe(3);
    expect(fetchMock).toHaveBeenCalledTimes(1);
    const url = fetchMock.mock.calls[0][0] as string;
    expect(url).toContain('/api/v2/runtime/prompts/refund-agent');
    expect(url).toContain('projectId=proj-1');
    expect(url).toContain('environment=production');
  });

  it('TTL 内直接返回缓存，不发请求', async () => {
    fetchMock.mockResolvedValueOnce(jsonResponse(SAMPLE));
    const client = makeClient();

    await client.get('refund-agent');
    await client.get('refund-agent');
    expect(fetchMock).toHaveBeenCalledTimes(1);
  });

  it('TTL 过期后带 If-None-Match 协商；304 时复用旧内容并刷新 TTL', async () => {
    vi.useFakeTimers();
    try {
      fetchMock
        .mockResolvedValueOnce(jsonResponse(SAMPLE, { headers: { ETag: SAMPLE.etag } }))
        .mockResolvedValueOnce(notModified(SAMPLE.etag));
      const client = makeClient();

      const first = await client.get('refund-agent');
      expect(first.promptVersionId).toBe('v-1');

      // 推进时钟超过 TTL
      vi.advanceTimersByTime(2000);

      const second = await client.get('refund-agent');
      expect(second.promptVersionId).toBe('v-1');
      expect(fetchMock).toHaveBeenCalledTimes(2);
      const secondCallHeaders = fetchMock.mock.calls[1][1].headers as Record<string, string>;
      expect(secondCallHeaders['If-None-Match']).toBe(SAMPLE.etag);

      // 再次未过期则不发请求
      await client.get('refund-agent');
      expect(fetchMock).toHaveBeenCalledTimes(2);
    } finally {
      vi.useRealTimers();
    }
  });

  it('200 响应覆盖旧缓存', async () => {
    fetchMock
      .mockResolvedValueOnce(jsonResponse(SAMPLE, { headers: { ETag: SAMPLE.etag } }))
      .mockResolvedValueOnce(
        jsonResponse(
          { ...SAMPLE, promptVersionId: 'v-2', versionNumber: 4, content: 'v4 content', etag: '"pv-new"' },
          { headers: { ETag: '"pv-new"' } }
        )
      );
    const client = makeClient();

    const v1 = await client.get('refund-agent');
    // 强制刷新
    const v2 = await client.get('refund-agent', { forceRefresh: true });
    expect(v1.promptVersionId).toBe('v-1');
    expect(v2.promptVersionId).toBe('v-2');
    expect(v2.content).toBe('v4 content');
  });

  it('不同环境走不同缓存键', async () => {
    const stagingSample = { ...SAMPLE, environment: 'staging' as const, etag: '"etag-stg"' };
    fetchMock
      .mockResolvedValueOnce(jsonResponse(SAMPLE, { headers: { ETag: SAMPLE.etag } }))
      .mockResolvedValueOnce(jsonResponse(stagingSample, { headers: { ETag: stagingSample.etag } }));
    const client = makeClient();

    const prod = await client.get('refund-agent', { environment: 'production' });
    const stg = await client.get('refund-agent', { environment: 'staging' });
    expect(prod.environment).toBe('production');
    expect(stg.environment).toBe('staging');
    expect(fetchMock).toHaveBeenCalledTimes(2);
    expect(client.cacheSize).toBe(2);
  });

  it('非 2xx/304 抛错', async () => {
    fetchMock.mockResolvedValueOnce(new Response('not found', { status: 404 }));
    const client = makeClient();
    await expect(client.get('missing')).rejects.toThrow(/404/);
  });

  it('clearCache 清空所有条目', async () => {
    fetchMock.mockResolvedValueOnce(jsonResponse(SAMPLE));
    const client = makeClient();
    await client.get('refund-agent');
    expect(client.cacheSize).toBe(1);
    client.clearCache();
    expect(client.cacheSize).toBe(0);
  });
});

File: sdk-go/agentmonitor/prompt_runtime_test.go
package agentmonitor

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"strings"
	"sync"
	"testing"
	"time"
)

const samplePromptJSON = `{
  "promptId":"p-1",
  "promptName":"refund-agent",
  "promptVersionId":"v-1",
  "versionNumber":3,
  "environment":"production",
  "content":"You are a refund agent v3",
  "variablesSchema":{"type":"object"},
  "modelDefaults":{"temperature":0.2},
  "etag":"\"pv-abc\"",
  "deployedAt":"2026-08-26T00:00:00Z"
}`

func newSampleServer(t *testing.T, handler http.HandlerFunc) (*httptest.Server, *PromptRuntimeClient) {
	t.Helper()
	srv := httptest.NewServer(handler)
	t.Cleanup(srv.Close)
	c := NewPromptRuntimeClient(srv.URL, "amt_test", "proj-1", 50*time.Millisecond, nil)
	return srv, c
}

func TestPromptRuntimeFirstFetch(t *testing.T) {
	var seenURL string
	var seenAuth string
	_, c := newSampleServer(t, func(w http.ResponseWriter, r *http.Request) {
		seenURL = r.URL.String()
		seenAuth = r.Header.Get("Authorization")
		w.Header().Set("ETag", `"pv-abc"`)
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(samplePromptJSON))
	})
	p, err := c.Get(context.Background(), "refund-agent", nil)
	if err != nil {
		t.Fatalf("Get: %v", err)
	}
	if p.PromptVersionID != "v-1" || p.VersionNumber != 3 {
		t.Fatalf("unexpected prompt: %+v", p)
	}
	if seenAuth != "Bearer amt_test" {
		t.Fatalf("expected bearer token, got %q", seenAuth)
	}
	if seenURL == "" || !strings.Contains(seenURL, "projectId=proj-1") || !strings.Contains(seenURL, "environment=production") {
		t.Fatalf("unexpected url: %s", seenURL)
	}
}

func TestPromptRuntimeTTLCacheHit(t *testing.T) {
	calls := 0
	_, c := newSampleServer(t, func(w http.ResponseWriter, r *http.Request) {
		calls++
		w.Header().Set("ETag", `"pv-abc"`)
		_, _ = w.Write([]byte(samplePromptJSON))
	})
	_, _ = c.Get(context.Background(), "refund-agent", nil)
	_, _ = c.Get(context.Background(), "refund-agent", nil)
	if calls != 1 {
		t.Fatalf("expected 1 network call within TTL, got %d", calls)
	}
}

func TestPromptRuntime304ExtendsTTL(t *testing.T) {
	var mu sync.Mutex
	calls := 0
	_, c := newSampleServer(t, func(w http.ResponseWriter, r *http.Request) {
		mu.Lock()
		calls++
		n := calls
		mu.Unlock()
		switch n {
		case 1:
			w.Header().Set("ETag", `"pv-abc"`)
			_, _ = w.Write([]byte(samplePromptJSON))
		case 2:
			if r.Header.Get("If-None-Match") != `"pv-abc"` {
				http.Error(w, "missing If-None-Match", http.StatusBadRequest)
				return
			}
			w.WriteHeader(http.StatusNotModified)
		default:
			http.Error(w, "should not be called again", http.StatusInternalServerError)
		}
	})
	first, err := c.Get(context.Background(), "refund-agent", nil)
	if err != nil {
		t.Fatalf("first Get: %v", err)
	}
	time.Sleep(80 * time.Millisecond) // 超过 TTL（50ms）
	second, err := c.Get(context.Background(), "refund-agent", nil)
	if err != nil {
		t.Fatalf("second Get: %v", err)
	}
	if first.PromptVersionID != second.PromptVersionID {
		t.Fatalf("304 should reuse cached prompt: first=%s second=%s", first.PromptVersionID, second.PromptVersionID)
	}
	// 304 应刷新 TTL，再次立即获取不应发请求
	_, _ = c.Get(context.Background(), "refund-agent", nil)
	mu.Lock()
	defer mu.Unlock()
	if calls != 2 {
		t.Fatalf("expected 2 calls, got %d", calls)
	}
}

func TestPromptRuntime200OverwritesCache(t *testing.T) {
	step := 0
	_, c := newSampleServer(t, func(w http.ResponseWriter, r *http.Request) {
		step++
		if step == 1 {
			w.Header().Set("ETag", `"pv-abc"`)
			_, _ = w.Write([]byte(samplePromptJSON))
			return
		}
		updated := ResolvedPrompt{
			PromptID:        "p-1",
			PromptName:      "refund-agent",
			PromptVersionID: "v-2",
			VersionNumber:   4,
			Environment:     "production",
			Content:         "v4",
			ETag:            `"pv-new"`,
		}
		w.Header().Set("ETag", `"pv-new"`)
		_ = json.NewEncoder(w).Encode(updated)
	})
	v1, _ := c.Get(context.Background(), "refund-agent", nil)
	v2, err := c.Get(context.Background(), "refund-agent", &GetPromptOptions{ForceRefresh: true})
	if err != nil {
		t.Fatalf("Get: %v", err)
	}
	if v1.PromptVersionID != "v-1" || v2.PromptVersionID != "v-2" || v2.Content != "v4" {
		t.Fatalf("unexpected versions v1=%+v v2=%+v", v1, v2)
	}
}

func TestPromptRuntimeEnvironmentIsolation(t *testing.T) {
	_, c := newSampleServer(t, func(w http.ResponseWriter, r *http.Request) {
		env := r.URL.Query().Get("environment")
		p := ResolvedPrompt{
			PromptID:        "p-1",
			PromptName:      "refund-agent",
			PromptVersionID: "v-" + env,
			VersionNumber:   1,
			Environment:     env,
			Content:         env,
			ETag:            `"etag-` + env + `"`,
		}
		_ = json.NewEncoder(w).Encode(p)
	})
	prod, _ := c.Get(context.Background(), "refund-agent", &GetPromptOptions{Environment: PromptEnvProduction})
	stg, _ := c.Get(context.Background(), "refund-agent", &GetPromptOptions{Environment: PromptEnvStaging})
	if prod.Environment != "production" || stg.Environment != "staging" {
		t.Fatalf("env mismatch: %+v vs %+v", prod, stg)
	}
	if c.CacheSize() != 2 {
		t.Fatalf("expected 2 cache entries, got %d", c.CacheSize())
	}
}

func TestPromptRuntimeErrorResponse(t *testing.T) {
	_, c := newSampleServer(t, func(w http.ResponseWriter, r *http.Request) {
		http.Error(w, "not found", http.StatusNotFound)
	})
	_, err := c.Get(context.Background(), "missing", nil)
	if err == nil || !strings.Contains(err.Error(), "404") {
		t.Fatalf("expected 404 error, got %v", err)
	}
}

func TestPromptRuntimeClearCache(t *testing.T) {
	_, c := newSampleServer(t, func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("ETag", `"pv-abc"`)
		_, _ = w.Write([]byte(samplePromptJSON))
	})
	_, _ = c.Get(context.Background(), "refund-agent", nil)
	if c.CacheSize() != 1 {
		t.Fatalf("expected cache size 1, got %d", c.CacheSize())
	}
	c.ClearCache()
	if c.CacheSize() != 0 {
		t.Fatalf("expected cache cleared, got %d", c.CacheSize())
	}
}
